### DISCUSSION TITLE
iOS app unable to sync

### DISCUSSION THREAD

**Discussion by zacknewman** [dani-garcia/vaultwarden#5065] [General]
Normally I would create a bug if I was _certain_ this was an issue with Vaultwarden; however my fork of Vaultwarden has deviated so much, that there is a non-negligible chance the issue is with my fork. If I were a betting man though—which I am—I'd wager money this issue affects Vaultwarden too.

Anyway, I integrated the recent changes made to 1.32.1; notably #4932 and #4937. The affected functions are _never_ called when logging in for the first time though which leads me to believe validation needs to occur other places. In particular `api::core::ciphers::sync` _is_ called, but there were no changes to it.

Wanted to start a discussion as a way to potentially share ideas. If there is definitely _not_ an issue with Vaultwarden, then the problem is mine alone so feel free to close this discussion. If this issue potentially affects Vaultwarden, then it would be nice to discuss which functions I should be looking for. Currently I'm spraying `println` in a bunch of places to see which functions do and do not get called.

**Comment by BlackDex** [comment]
If anything, it might be linked to #4932.
But that is only run during sync when the ciphers are converted from there struct to json representation.

Best to start with is checking if you see anything out of the ordinary in the response of the sync call. Doesn't matter if it is from the web-vault or via your ios device. The result should be the same. 

And see if you have any strange passwordhistory items even after the fixes.

I would really love to know if you find anything. Because i have not yet been able to reproduce this my self, or see anything strange yet in files sent to me by users.

**Comment by zacknewman** [comment]
As stated, I'm using a lot of `println` calls; so I already have the JSON `api::core::ciphers::sync` sends to the client; however it's not obvious what is considered "out of the ordinary". There is certainly a lot of `null` values, but I don't know which are problematic.

Do you happen to know the typical sequence of API calls? As stated, `api::core::ciphers::sync` is called; however I'd like to know what the next function calls would be to narrow down the last API call as presumably it is the problem. Obviously I can read the source code for the iOS app, but I was hoping for a "shortcut".

**Reply by BlackDex** [reply]
I have no clue what would be "out of the ordinary", if i would know that on sight by mind then I would have solved it already.

As far as i know, the `/api/sync` is called, and that is where the client stops, no other calls are done.
The client starts to process the json and decode the values etc.. etc... And somewhere there it breaks, some value is causing an issue. But which value is not clear. The Android clients also do not provide a detailed log message.

I am currently still trying to shift through sync's send to me to try and debug this. And i tried several ways to break stuff already. But I'm not able to. Mostly modifying the date/time for example, or adding `null`'s where they shouldn't.

**Reply by zacknewman** [reply]
> As far as i know, the /api/sync is called, and that is where the client stops, no other calls are done.

OK, then I will dedicate time to `api::core::ciphers::sync`.

> The Android clients also do not provide a detailed log message.

My wife is not having issues, so I assumed this was an iOS only thing. The Firefox web extension works fine as does the web vault. I'll log into my wife's phone to see if I experience a problem with it.

**Reply by zacknewman** [reply]
Android app works fine; however the Android app still have not been updated as 2024.7.1 is the latest version in the app store.

**Reply by BlackDex** [reply]
Ok, cool.

It could be as simple as date/time having too many or too little nano-seconds for example.
Or even a profile item maybe.

But it must be somewhere in the response of the sync, which of course does include all the collections, folders, organizations, the users profile and all the ciphers of course.

**Reply by zacknewman** [reply]
Had my wife log into my phone, and it messes up; so it's not isolated to my password vault. Also WebAuthn 2FA and TOTP 2FA no longer work. I removed all 2FA to see if that was the only issue, but clearly with the vault sync, it is not. At first I thought it was just a WebAuthn thing; however my wife uses TOTP, and it fails too. Again this is on the iOS app; however I suspect if/when Android gets updated, it will affect it as well. For now, I told my wife to disable auto-updates.

**Reply by BlackDex** [reply]
You can download GitHub builds here https://github.com/bitwarden/android/actions/workflows/build.yml

I have tried several versions already and not been able to see any errors as of yet.

I even installed the one build like an hour ago, works just fine.

**Reply by BlackDex** [reply]
There were several changes for 2FA done also. Not sure if you backported those too? #4762 for example.

But regarding the 2FA, are there any errors on the Vaultwarden side? Maybe some differences in json key cases still somewhere?

**Reply by zacknewman** [reply]
I'll verify, but my fork should be up-to-date with 1.32.1. I only have code for TOTP and WebAuthn as I prefer to not include code I don't want—one reason why it's not correct of me to create an issue since my fork can hardly be called "Vaultwarden". I'll double check that the JSON that is sent for TOTP and WebAuthn is the same.

> But regarding the 2FA, are there any errors on the Vaultwarden side? Maybe some differences in json key cases still somewhere?

I've only spent time trying to figure out the sync issues, but I'll quickly check now about the 2FA. To emphasize again, the current Android app, Firefox extension, and web vault don't have issues syncing or with WebAuthn and TOTP 2FA. These issues only affect the current iOS app. I can't recall the last time I used the iOS app before today, but I imagine it's been at most 2 or 3 weeks and it was fine then.

**Reply by zacknewman** [reply]
For WebAuthn, the authentication challenge is created (i.e., `api::core::two_factor::webauthn::generate_webauthn_login` is successfully called); but the ceremony is never completed (i.e., `api::core::two_factor::webauthn::validate_webauthn_login` is never called). `api::identity::twofactor_auth` is only called once. I wonder if Bitwarden finally fixed the "hack" where an error is (ab)used to indicate 2FA forcing clients to re-try.

Specifically, the way things are currently designed, `twofactor_auth` should be called exactly twice when 2FA is enabled. First time causes `_json_err_twofactor` to be called which errors and informs the client that 2FA is enabled and what forms are enabled. Second time `twofactor_auth` is called `data.two_factor_token` is populated allowing the WebAuthn ceremony to complete.

Perhaps it's something else. Seeing how `twofactor_auth` is not called twice though, I'm guessing the JSON error that is returned after the first call is the problem. Weird that the Firefox extension works though seeing how it's been updated too.

**Reply by BlackDex** [reply]
So, on my Android i have no issues at all regarding this. Both WebAuthn and TOTP.
I was using version: 2024.10.0 (19290) which i downloaded from here https://github.com/bitwarden/android/actions/runs/11263511400

So, i doubt that they have _fixed_ that.
It could be the JSON which your Vaultwarden returns and causes an issue.

As a reference it should look something like this (I have all 2FA options enabled and configured):
Also note the case of the keys here, those are important.
```json
{
    "MasterPasswordPolicy": {
        "Object": "masterPasswordPolicy"
    },
    "TwoFactorProviders": [
        "0",
        "1",
        "2",
        "3",
        "7"
    ],
    "TwoFactorProviders2": {
        "0": null,
        "1": {
            "Email": "vw*****@domain.tld"
        },
        "2": {
            "AuthUrl": "https://api-4f2dbfa5.duosecurity.com/oauth/v1/authorize?response_type=code&client_id=D...J&request=eyJ---.eyJ---"
        },
        "3": {
            "Nfc": true
        },
        "7": {
            "allowCredentials": [
                {
                    "id": "9h...-u...5-O...g",
                    "type": "public-key"
                },
                {
                    "id": "B7...-...vQ",
                    "type": "public-key"
                }
            ],
            "challenge": "pQ...-...uD-...bw",
            "extensions": {
                "appid": "https://vw.domain.tld/app-id.json",
                "getCredBlob": false
            },
            "rpId": "vw.domain.tld",
            "timeout": 60000,
            "userVerification": "discouraged"
        }
    },
    "error": "invalid_grant",
    "error_description": "Two factor required."
}
```

**Reply by zacknewman** [reply]
Ugh, what an idiot. I never changed `TwoFactorProviders` to be an array of strings; instead it was an array of integers. Now both TOTP and WebAuthn 2FA work. Sorry for this. I _think_ there is still an issue with sync, but clearly I cannot be trusted that I didn't forget something on my side.

If I happen to figure out how to fix sync, I'll report back; and perhaps it can help you. Until then, just ignore this discussion as it could very well be isolated to my fork. Thanks for the help.

**Reply by BlackDex** [reply]
I have an iOS device now. But my first attempt to sync with my private vault worked just fine.
Now going to see if i can reproduce the same on my dev environment, but I'm afraid it will be difficult.

Any feedback would be great! Thanks!

**Comment by BlackDex** [comment]
@zacknewman, how is your SSL configured for the app?
Could you maybe use some local tools like [sslyze](https://github.com/nabla-c0d3/sslyze) or [testssl.sh](https://github.com/drwetter/testssl.sh) and check if there are any SSL issues?

Both tools should show a list of compatibility with clients like iOS for example.
While i do not think that would be an issue, since  you are able to login, who knows what further down the line could be causing an issue.

**Reply by zacknewman** [reply]
TLS is not the issue. The issue is folders. As soon as I moved everything from a folder to "no folder", I can sync; furthermore, I get an error when I try to create a folder. Also how does one create a folder in the web vault? I don't create folders often, but I'm clicking on everything and don't see an option. I do have the option in the iOS app; and as stated, it errors.

**Reply by BlackDex** [reply]
You get an error when you create a folder via the iOS app?
Because i have several accounts I tested which have folders, and none broke.

How does the folder json look like?

**Reply by zacknewman** [reply]
If there exists even one folder, the iOS app is unable to sync. Once there are no folders, it can which luckily rules out the individual items since I didn't remove a single one in my vault nor the organization I am in. After successfully syncing, when I add a folder, I get an error which does not surprise me since folders were the issue that prevented the app from syncing.

How does one add a folder in the web vault?

> How does the folder json look like?

I'll get back to you on that. I spent 30 minutes figuring out the issue is related to folders, but I haven't spent time on actually finding the issue and how to fix it. Now that I know it's folder related (for me in my "fork), I can explore further.

**Reply by BlackDex** [reply]
Strange. I'm not able to reproduce this at all on the iOS device i have.

To create a folder via the mobile client do the following.
- Settings
- Vault
- Folders
- Click on the `+` sign at the top right.

**Reply by zacknewman** [reply]
No, I know how to create a folder in the iOS app since as stated it causes an error. I want to know how to create a folder in the _web vault_? I am clicking everywhere, and I don't see any "add folder option".

**Reply by BlackDex** [reply]
Ah, sorry!
It's as simple as clicking new at the top right.
![Screenshot_20241010-182635~2](https://github.com/user-attachments/assets/a2ad3c8b-918c-423e-a23a-6a0567650b32)

**Reply by zacknewman** [reply]
Ah, yes. I forgot that was how. OK, I went ahead and created an empty folder in the web vault; and that was enough to cause sync issues with the iOS client (i.e., _any_ folder (even an empty one) causes problems).

**Reply by BlackDex** [reply]
Doesn't make it more clear, as I'm unable to reproduce that my self unfortunately.

**Reply by zacknewman** [reply]
I understand. I'm merely reporting information in the off chance it helps you. Feel free to ignore this discussion as it may be isolated to me and my fork. If it's inappropriate for me to say anything, then please say so and I'll stop this thread.

**Reply by BlackDex** [reply]
Ow no, im also just stating my frustrations about being unable to see the exact same issue and also other users seem to have this issue, but not being able to pinpoint it.

**Reply by zacknewman** [reply]
Honestly unless you read the source code for the iOS app or a user is willing to send you privately their actual database and master password (which would be insane), I don't see how you will be able to fix this. I'm "lucky" in that I have an actual vault that causes issues, so at least I have something to work with. The fact you're able to successfully create a folder makes me less confident my problem is also a Vaultwarden one; at the same time, the fact other users of Vaultwarden proper are encountering issues, makes me think there is a _chance_ the issue is shared.

It could be database related as well. My fork is not only different at the application level, but my SQLite database tables are also different. That would perhaps also explain why some users of Vaultwarden proper are having problems since they may have not updated their database properly? Your guess is as good as mine.

**Reply by BlackDex** [reply]
I wouldn't need the password. Just the sync data should be enough, and a few people already did send me those, also `har` files, just not regarding this case yet. Though, even with that data, I'm not able to reproduce those users issues sometimes, but it does help me.

In theory the database should be migrated automatically, and if that fails it will produce an error, so that shouldn't happen. But a corrupted database could still cause strange issues of course.

I'm still curious into what you will find (and how you were able to find that), it might help with future issues.

**Reply by zacknewman** [reply]
> How does the folder json look like?

```json
"folders":[
    {
      "Id":"a50a8b63-155c-4794-abe5-d09af7059c5d",
      "Name":"2.2Efey7x2X1rtfDINy3chMw==|ZWYTva6PZjmXCL92GvEj/w==|4ioywLsbefbMchyhfFm5TE9jJmKHSjIAgi1HYlhB48E=",
      "Object":"folder",
      "RevisionDate":"2021-10-19T00:44:33.582304Z"
    },
    ⋮
]
```

That's what the the `folders` JSON array looks like.

P. S. I'm very careful about what I choose to reveal and not reveal; so while others may not want to reveal their e-mail address, domain, IP address, user ID, folder ID, WebAuthn Credential ID, etc.; you can be confident that information I do reveal is allowed within my threat model (i.e., don't "panic" and scrub data you would from my message).

**Reply by BlackDex** [reply]
I think you need to convert it to camelCase, as you now have PascalCase for these items.

In Vaultwarden we convert everything to camelCase.

https://github.com/dani-garcia/vaultwarden/blob/f0efec7c9623fab4ba3c2b53cba724e4e6a8a739/src/db/models/folder.rs#L42...L52

And this goes for almost all output, but some are still using PascalCase.

**Reply by zacknewman** [reply]
Yeah, once I sent that that stood out. Ugh, I thought I converted all the PascalCase to camelCase already. I'll have to properly vet all of the JSON to see if there is somewhere else I missed.

**Reply by zacknewman** [reply]
FYI,

Vaultwarden has a typo in how [`TwoFactor` is serialized](https://github.com/dani-garcia/vaultwarden/blob/f0efec7c9623fab4ba3c2b53cba724e4e6a8a739/src/db/models/two_factor.rs#L59) (`Oobject` instead of `object`).

Also [the JSON you sent](https://github.com/dani-garcia/vaultwarden/discussions/5065#discussioncomment-10902208) for the two factor stuff is still PascalCase (e.g., `MasterPasswordPolicy`). Mine is too, and the iOS app worked. Is that intentional?

**Reply by BlackDex** [reply]
Ah, thanks for the typo notice.

And strangely yes. It's either something Bitwarden forgot, or they need it for something else in the PascalCase format.
But it is both on the Android and iOS client. And the Bitwarden SaaS also still sends it this way. 

Converting it to camelCase will break the mobile clients.

**Reply by zacknewman** [reply]
OK, there were a couple other places that I somehow missed; anyway, I am able to successfully sync now. Thanks for your time. I apologize for starting this discussion as it's one's responsibility to fix their own self-inflicted issues. I'll exercise better judgement next time before hastily starting a discussion. I foolishly convinced myself because I saw other people having issues that it was a shared one. Again, thank you.

**Reply by BlackDex** [reply]
np. Sometimes it's good to reflect. And, you found a type in the process.
